include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,340 @@
+// Copyright (C) 2018-2025 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_BXDF_COOK_TORRANCE_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BXDF_COOK_TORRANCE_INCLUDED_
+
+#include "nbl/builtin/hlsl/bxdf/common.hlsl"
+#include "nbl/builtin/hlsl/bxdf/config.hlsl"
+#include "nbl/builtin/hlsl/bxdf/ndf.hlsl"
+#include "nbl/builtin/hlsl/bxdf/fresnel.hlsl"
+#include "nbl/builtin/hlsl/bxdf/ndf/microfacet_to_light_transform.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace bxdf
+{
+
+namespace impl
+{
+template<typename T, typename U>
+struct __implicit_promote;
+
+template<typename T>
+struct __implicit_promote<T,T>
+{
+    static T __call(const T v)
+    {
+        return v;
+    }
+};
+
+template<typename T>
+struct __implicit_promote<T,vector<typename vector_traits<T>::scalar_type, 1> >
+{
+    static T __call(const vector<typename vector_traits<T>::scalar_type, 1> v)
+    {
+        return hlsl::promote<T>(v[0]);
+    }
+};
+
+template<class N, class F, bool IsBSDF>
+struct quant_query_helper;
+
+template<class N, class F>
+struct quant_query_helper<N, F, true>
+{
+    using quant_query_type = typename N::quant_query_type;
+
+    template<class C>
+    static quant_query_type __call(NBL_REF_ARG(N) ndf, NBL_CONST_REF_ARG(F) fresnel, NBL_CONST_REF_ARG(C) cache)
+    {
+        return ndf.template createQuantQuery<C>(cache, fresnel.orientedEta.value[0]);
+    }
+};
+
+template<class N, class F>
+struct quant_query_helper<N, F, false>
+{
+    using quant_query_type = typename N::quant_query_type;
+
+    template<class C>
+    static quant_query_type __call(NBL_REF_ARG(N) ndf, NBL_CONST_REF_ARG(F) fresnel, NBL_CONST_REF_ARG(C) cache)
+    {
+        typename N::scalar_type dummy;
+        return ndf.template createQuantQuery<C>(cache, dummy);
+    }
+};
+
+template<class F, bool IsBSDF>
+struct check_TIR_helper;
+
+template<class F>
+struct check_TIR_helper<F, false>
+{
+    template<class MicrofacetCache>
+    static bool __call(NBL_CONST_REF_ARG(F) fresnel, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        return true;
+    }
+};
+
+template<class F>
+struct check_TIR_helper<F, true>
+{
+    template<class MicrofacetCache>
+    static bool __call(NBL_CONST_REF_ARG(F) fresnel, NBL_CONST_REF_ARG(MicrofacetCache) cache)
+    {
+        return cache.isValid(fresnel.getRefractionOrientedEta());
+    }
+};
```

Comment:
I think its possible to move these into `NBL_IUF_CONSTEXPR` if you just made a struct which just concerns itself with 
```cpp
scalar_type orientedEta = impl::get_orientedEta_helper(fresnel); // return `fresnel.getRefractionOrientedEta()` if `SupportsTransmission` requiring a twosided fresnel, otherwise return uninitialized var
```

Reply:
`check_TIR_helper` not needed anymore actually.